### PR TITLE
Wrap render method created using class properties (2)

### PR DESCRIPTION
**Pull request by nkzawa**
Fixes https://github.com/zeit/next.js/issues/828

### CODE REVIEW

File: client/patch-react.js [hunk 1]
```diff
@@ -17,8 +17,23 @@ export default (handleError = () => {}) => {
   React.createElement = function (Component, ...rest) {
     if (typeof Component === 'function') {
       const { prototype } = Component
-      if (prototype && prototype.render) {
-        prototype.render = wrapRender(prototype.render)
+
+      // assumes it's a class component if render method exists.
+      const isClassComponent = !!(prototype && prototype.render) ||
```
Comment (arunoda) [inline client/patch-react.js]: Using `!!` is seems pretty weird. Either don't use them or use `Boolean(xxx)`

File: client/patch-react.js [hunk 2]
```diff
@@ -17,8 +17,23 @@ export default (handleError = () => {}) => {
...
+        // subclass of React.Component or PureComponent with no render method.
+        prototype instanceof React.Component ||
+        prototype instanceof React.PureComponent
+
+      if (isClassComponent) {
+        // There's no render method in prototype
```
Comment (arunoda) [inline client/patch-react.js]: Is this comment correct? (or in the correct place?)

**Comment by nkzawa** [comment]
Fixed!

File: client/patch-react.js [hunk 3]
```diff
@@ -17,8 +17,23 @@ export default (handleError = () => {}) => {
...
-      if (prototype && prototype.render) {
-        prototype.render = wrapRender(prototype.render)
+
+      // assumes it's a class component if render method exists.
+      const isClassComponent = Boolean(prototype && prototype.render) ||
+        // subclass of React.Component or PureComponent with no render method.
+        // There's no render method in prototype
+        // when it's created with class-properties.
+        prototype instanceof React.Component ||
+        prototype instanceof React.PureComponent
```
Comment (rauchg) [inline client/patch-react.js]: Will this not make it work with class properties and alt-reacts like Preact? cc @developit

Comment (rauchg) [inline client/patch-react.js]: I guess the `React` compat could make the `instanceof` pass, right?

Comment (developit) [inline client/patch-react.js]: The alias will make this work, yup.

**Pull request merged**